teammapper-frontend/mmp/src/map/handlers/nodes.ts

```diff
@@ -664,68 +689,141 @@ export default class Nodes {
     }
 
     /**
-     * Return the appropriate coordinates of the node.
-     * @param {Node} node
-     * @returns {Coordinates} coordinates
+     * Base method for calculating node coordinates. 
+     * The reason this exists is so we can work with a JSON snapshot (as given by an import), but also allow saved, "real" nodes to calculate coordinates
+     * This prevents duplication, whilst passing methods that differ depending on whether or not a JSON snapshot or "real" node is calculating coordinates.
+     * @param node Either a node previously saved or one from a JSON snapshot
+     * @param params
+     * getParent() - Method to find the parent of the given node
+     * getSiblings() - Method to get the siblings of the given node
+     * isRoot() - If parent node is root
+     * getOrientation() - Method to get the orientation of the node
+     * @returns 
      */
-    private calculateCoordinates(node: Node): Coordinates {
+    private calculateNodeCoordinates(
+        node: Node | ExportNodeProperties,
+        params: {
+            getParent: () => (Node | ExportNodeProperties) | null,
```

Comment:
do these really need to be functions and not just of type node or something? I think in both cases (only checked getParent and getSiblings) when you are calling this function, you are passing in a lambda that simply calculates a bit later (when it is called):

E.g.
```
getParent: () => node.parent
```

```
const nodeParent = params.getParent();
```

Here, you could simply pass in the parent as a node instead of using a lambda.

Reply:
Yeah I think for `getParent` it makes sense, but the other arguments are actually methods (`this.getSiblings()` for example), so it makes sense to keep the lambda there.

Reply:
I've changed `getParent` to `nodeParent` and it's no longer a lambda.

Reply:
Just to clarify regarding `getSiblings`: In the case of a JSON import, it's simply calling `map.filter()`, which is not exactly a method, but in the case of real nodes, it calls `this.getSiblings()` in `nodes.ts`, which requires proper `this` context. `getOrientation` actually calls a method in both cases, hence why I think it makes sense to keep both lambdas there.